lib/src/merge.rs

```diff
@@ -296,15 +305,23 @@ impl<T> IntoIterator for Merge<T> {
 
 impl<T> FromIterator<T> for MergeBuilder<T> {
     fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
-        let mut removes = vec![];
-        let mut adds = vec![];
-        let mut curr = &mut adds;
-        let mut next = &mut removes;
+        let mut builder = MergeBuilder::default();
+        builder.extend(iter);
+        builder
+    }
+}
+
+impl<T> Extend<T> for MergeBuilder<T> {
+    fn extend<I: IntoIterator<Item = T>>(&mut self, iter: I) {
+        let (mut curr, mut next) = if self.adds.len() == self.removes.len() + 1 {
```

Comment:
nit: I slightly prefer `self.adds.len() != self.removes.len()` because it looks more general than `x == y + 1`, but that's just a matter of taste.